### ISSUE TITLE
Newznab provider broken

### ISSUE THREAD

**Issue by vajonam** [SickGear/SickGear#586]
Looks like the newznab provider is broken. Where the newznab.py is sending the "starifi-ed" api key. This is then resulting in HTML which cannot be parsed with XML. 

I looked for a fix so I coud create a PR to send but having just started looking at the code. in the config.ini the API key is stored correctly 

I added some debuggin in newznab.py to print out what we are getting, but looks like we are already getting the only the starrred key. 

```
2015-12-09 01:03:55 DEBUG    TORNADO :: APIKEY  = 86d3************************b324
2015-12-09 01:03:55 DEBUG    TORNADO :: Sending Request to [http://dvr/spotweb/api?apikey=86d3************************b324&t=caps]
2015-12-09 01:03:55 DEBUG    TORNADO :: Error trying to parse xml data. Error: mismatched tag: line 49, column 78Data: <html>
	<head>
		<style type='text/css'>
			html {
				overflow:auto;
				height:100%;
			}

			body {
				margin: 10px;
				padding: 25px 0 0 0;
				font:11px Arial, Helvetica, sans-serif;
			}

			div.permdenied {
				border: 1px solid red;
				background-color: #fef1fc;
				color: #cd0a0a;

				font-size: 0.9em;
				
				margin-top: 4px;
				margin-left: auto;
				margin-right: auto;
				
				border-top-left-radius: 4px;
				border-top-right-radius: 4px;
				border-bottom-right-radius: 4px;
				border-bottom-left-radius: 4px;
			} 
			
			div.permdenied p {
				text-align: center;
			} 

			ul.formerrors {
				padding: 10px;
				color: #cd0a0a;
				list-style: none;
			}

					</style>
	</head>

	<body>
		<div class='container'>
			<div class='permdenied'>
				<p>
					Access denied for [Use Spotweb using an API key (24::invalid API key)] </strong>
				</p>
			</div>
		</div>
	
	<br>
	
	<ul class='formerrors'></ul><ul class='forminformation'></ul><form class="loginform" name="loginform" action="/spotweb/?page=login" method="post">
	<input type="hidden" name="loginform[xsrfid]" value="1449641035:loginform:0:1e6e929034f8af9ebc16cb6aca2ce393e763a3a1">
	<input type="hidden" name="loginform[http_referer]" value="http://dvr/spotweb/?page=newznabapi&amp;apikey=86d3%2A%2A%2A%2A%2A%2A%2A%2A%2A%2A%2A%2A%2A%2A%2A%2A%2A%2A%2A%2A%2A%2A%2A%2Ab324&amp;t=caps">
			<input type="hidden" name="data[performredirect]" value="1">
				<input type="hidden" name="data[renderhtml]" value="1">
		<fieldset>
		<dl>
			<dt><label for="loginform[username]">Username</label></dt>
			<dd><input type="text" name="loginform[username]" value=""></dd>

			<dt><label for="loginform[password]">Password</label></dt>
			<dd><input type="password" name="loginform[password]" value=""></dd>

			<dd><input class="greyButton" type="submit" name="loginform[submitlogin]" value="Login"></dd>
		</dl>
	</fieldset>
</form>
</div></body></html>
2015-12-09 01:03:55 DEBUG    TORNADO :: Error parsing xml for [mspotweb]
```

**Comment by adam111316** [comment]
You need to provide some more information, eg branch / version
https://github.com/SickGear/SickGear/wiki/%5BHow-to%5D-Report-Issues

**Comment by adam111316** [comment]
Also I don't believe there is any code in newznab.py that would cause starring. So if anything a better place to look would be where the apikey is being saved / loaded, aka webserve.py or maybe the main __init__.py script

**Comment by vajonam** [comment]
@adam111316 sorry, this against the both the develop branch and as well the release/0.11.0 branch. I will have a look at the init.py where its being loaded.

**Comment by vajonam** [comment]
@adam111316 so I think know the problem, the problem is that we have array of newznab settings for each newznab provider, and we load it into the JS with the stars, so when the text method is executed the wrong key is sent. 

this will ix it but gets rid of the hashing and reduces security of being able to get the API keys from the UI. Not sure what an elegant fix, is sort of having getNewznabCategories method use the config settings from memory that what is passed in from the UI. 

```
diff --git a/gui/slick/interfaces/default/config_providers.tmpl b/gui/slick/interfaces/default/config_providers.tmpl
index f2f2268..cc99056 100644
--- a/gui/slick/interfaces/default/config_providers.tmpl
+++ b/gui/slick/interfaces/default/config_providers.tmpl
@@ -39,7 +39,7 @@
 
         #for $curNewznabProvider in $sickbeard.newznabProviderList:
 
-    \$(this).addProvider('$curNewznabProvider.get_id()', '$curNewznabProvider.name', '$curNewznabProvider.url', '<%= starify(curNewznabProvider.key) %
+    \$(this).addProvider('$curNewznabProvider.get_id()', '$curNewznabProvider.name', '$curNewznabProvider.url', '$curNewznabProvider.key', '$curNewzna
 
         #end for
```

**Comment by vajonam** [comment]
Another potential solution is to do what the testSABnzbd method does where it gets the key. How can I get the key similarly for newznab providers? 

```
        if None is not apikey and starify(apikey, True):
            apikey = sickbeard.SAB_APIKEY
```

**Comment by adam111316** [comment]
Yeah definitely wouldn't be looking at removing the starify as that's just a 'quick hack'.

Haven't really looked too hard at it, but it looks like the current code just creates a temp provider with whats in the form. eg:
`tempProvider= newznab.NewznabProvider(name, url, key)`

What I would be doing is pulling out the current saved newznab provider key and comparing it against what the form has (comparing both keys starred ofc). If its different use the form key, else use the saved key.

There seems to be a bit of commented code there too which could be used as a starting point:
`#providerDict = dict(zip([x.get_id() for x in sickbeard.newznabProviderList], sickbeard.newznabProviderList))`

Basically need to filter our provider list for the newznab provider name. Then just do newznabprovidername.key

**Comment by adam111316** [comment]
Just had a check through, it appears the 't=caps' statement does not require an api_key to work.

Even just pasting that URL into a browser does not work. You'll need to configure your spotweb program correctly.

**Comment by adam111316** [comment]
Appears some spotweb / NN sites bitch about no apikey / invalid apikeys being sent regardless if they are required or not. Patch for that will be coming shortly, however it still stands that your spotweb does not appear to be returning any correct responses for those urls. (What I get is different to what you've posted)

**Comment by vajonam** [comment]
Don't know how valid that is. Because it works fine on when I hardcore the API key in the key. Works with sickbeard and SickRage and other stuff. Not sure what you are seeing. For spotweb to work with newznab API you need add some URL rewrite rules as part of the install otherwise this won't work.

**Comment by vajonam** [comment]
Also the issue remains that we are passing invalid API keys as part of the temp provider. Not sure why the issue is invalid. Was trying to help.

**Comment by vajonam** [comment]
My spotweb isn't available to the internet. Now sure what you mean by "your spotweb doesn't seem to be returning the correct response for these URLs"

https://github.com/spotweb/spotweb/wiki/Spotweb-als-Newznab-Provider

Works fine for couchpotato, SickRage and sickbeard.

**Comment by adam111316** [comment]
Ah, I didn't realize that it wasn't available to the net, hence when I was doing my tests, it looked like it wasn't working.

Anyway as I said, there is a patch coming for this.

**Comment by adam111316** [comment]
Refer #588

**Comment by JackDandy** [comment]
@vajonam thanks for the report.

**Comment by JackDandy** [comment]
#588
